### ISSUE TITLE
False positive mismatched types [E0308]

### ISSUE THREAD

**Issue by imkow** [intellij-rust/intellij-rust#8047]
## Environment

* **IntelliJ Rust plugin version:** 0.4.160.4220-212-nightly
* **Rust toolchain version:** 1.58.0-nightly (18bc4bee9 2021-11-02) x86_64-pc-windows-msvc
* **IDE name and version:** CLion 2021.2.3 (CL-212.5457.51)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
https://github.com/google/tarpc/blob/master/plugins/src/lib.rs#L291

```rust
        return_types: &rpcs
            .iter()
            .map(|rpc| match rpc.output {
                ReturnType::Type(_, ref ty) => ty,
                ReturnType::Default => unit_type,
            })
            .collect::<Vec<_>>(),
```
